src/main/osd/osd_warnings.c

```diff
@@ -325,54 +411,41 @@ void renderOsdWarning(char *warningText, bool *blinking, uint8_t *displayAttr)
 #if defined(USE_DSHOT) && defined(USE_DSHOT_TELEMETRY)
     // Show esc error
     if (motorConfig()->dev.useDshotTelemetry && osdWarnGetState(OSD_WARNING_ESC_FAIL) && ARMING_FLAG(ARMED)) {
-        uint32_t dshotEscErrorLengthMotorBegin;
-        uint32_t dshotEscErrorLength = 0;
-
-        // Write 'ESC'
-        warningText[dshotEscErrorLength++] = 'E';
-        warningText[dshotEscErrorLength++] = 'S';
-        warningText[dshotEscErrorLength++] = 'C';
-
-        for (uint8_t k = 0; k < getMotorCount(); k++) {
-            // Remember text index before writing warnings
-            dshotEscErrorLengthMotorBegin = dshotEscErrorLength;
-
-            // Write ESC nr
-            warningText[dshotEscErrorLength++] = ' ';
-            warningText[dshotEscErrorLength++] = '0' + k + 1;
-
-            // Add esc warnings
-            if (osdConfig()->esc_rpm_alarm != ESC_RPM_ALARM_OFF
-                    && isDshotMotorTelemetryActive(k)
-                    && (dshotTelemetryState.motorState[k].telemetryData[DSHOT_TELEMETRY_TYPE_eRPM] * 100 * 2 / motorConfig()->motorPoleCount) <= osdConfig()->esc_rpm_alarm) {
-                warningText[dshotEscErrorLength++] = 'R';
-            }
+        uint8_t dshotEscErrorLength = 0;
+        bool escWarning = false;
 
-            // Skip if no extended telemetry available
-            bool edt = (dshotTelemetryState.motorState[k].telemetryTypes & DSHOT_EXTENDED_TELEMETRY_MASK) != 0;
-            
-            if (edt && osdConfig()->esc_temp_alarm != ESC_TEMP_ALARM_OFF
-                    && (dshotTelemetryState.motorState[k].telemetryTypes & (1 << DSHOT_TELEMETRY_TYPE_TEMPERATURE)) != 0
-                    && dshotTelemetryState.motorState[k].telemetryData[DSHOT_TELEMETRY_TYPE_TEMPERATURE] >= osdConfig()->esc_temp_alarm) {
-                warningText[dshotEscErrorLength++] = 'T';
-            }
-            if (edt && osdConfig()->esc_current_alarm != ESC_CURRENT_ALARM_OFF
-                    && (dshotTelemetryState.motorState[k].telemetryTypes & (1 << DSHOT_TELEMETRY_TYPE_CURRENT)) != 0
-                    && dshotTelemetryState.motorState[k].telemetryData[DSHOT_TELEMETRY_TYPE_CURRENT] >= osdConfig()->esc_current_alarm) {
-                warningText[dshotEscErrorLength++] = 'C';
+        // Write 'ESC' prefix
+        dshotEscErrorLength += tfp_sprintf(warningText + dshotEscErrorLength, "ESC");
+
+        for (unsigned i = 0; i < getMotorCount(); i++) {
+            uint8_t dshotEscErrorLengthMotorBegin = dshotEscErrorLength;
+
+            // Write ESC nr with proper multi-digit support using tfp_sprintf
+            dshotEscErrorLength += tfp_sprintf(warningText + dshotEscErrorLength, " %d", i + 1);
+
+            if (isDshotMotorTelemetryActive(i)) {
+                escSensorData_t escData;
+                if (getDshotSensorData(i, &escData) == 0) {
```

Comment:
(IMO is is slightly better to use <0 for error, >= 0 ok (with == 0 sometimes meaning 'no operation')
```suggestion
                if (getDshotSensorData(i, &escData) >= 0) {
```